122 documentation website mkdocs (#192)
eScienceLab/Cratey-Validator: docs/api.md

# API Reference

The examples below use the Compose stack's local address, `http://localhost:5001`. Note that the service serves its own OpenAPI specification at `/docs`.

!!! note

>     `POST /v1/ro_crates/validate_metadata` is always available, but the **storage-backed endpoints** are only available when the service runs with `STORAGE_ENABLED=true` (see [Installation & Setup](installation.md#enabling-object-storage)); without this set, requests return `404`. 

Author: ```suggestion
    `POST /v1/ro_crates/validate_metadata` is always available, but the **storage-backed endpoints** are only available when the service runs with `STORAGE_ENABLED=true` (see [Installation & Setup](installation.md#enabling-object-storage)); without this set, `POST /v1/ro_crates/{crate_id}/validation` or `GET /v1/ro_crates/{crate_id}/validation` requests will return a `404` result. 
```

## Validate metadata

`POST /v1/ro_crates/validate_metadata`

This validates the contents of an `ro-crate-metadata.json` document and returns the result in the response.

| Field | Required | Description |
|-------|----------|-------------|
| `crate_json` | yes | The metadata document, as a JSON string |

> | `profile_name` | no | Profile to validate against, e.g. `ro-crate-1.2`; defaults to `ro-crate-1.1` when omitted |

Author: ```suggestion
| `profile_name` | no | Profile to validate against, e.g. `ro-crate-1.2`. The validator will default to `ro-crate-1.1` when this is omitted |
```

!!! warning

>     Currently, the validation profile is not detected from the RO-Crate. In other words, a `conformsTo` declaration in the metadata does not influence which validation profile is used by the validator, and the validation always runs against `profile_name`, or `ro-crate-1.1` when it is omitted.

Author: Is this a problem with our code or with the validator?

> To validate a file:

Author: ```suggestion
To validate an ro-crate metadata file:
```

```bash
jq -Rs '{crate_json: .}' ro-crate-metadata.json | curl -X POST http://localhost:5001/v1/ro_crates/validate_metadata -H 'Content-Type: application/json' -d @-
```

> To choose a profile, add it to the `jq` object: `jq -Rs '{crate_json: ., profile_name: "ro-crate-1.2"}' ro-crate-metadata.json`.

Author: ```suggestion
To choose a profile, add it to the json object as a `profile_name` entry:
\```
jq -Rs '{crate_json: ., profile_name: "ro-crate-1.2"}' ro-crate-metadata.json | curl -X POST http://localhost:5001/v1/ro_crates/validate_metadata -H 'Content-Type: application/json' -d @-
\```
```

| Code | Meaning |
|------|---------|
| `200` | Validated; the result has a `status` of `valid` or `invalid` |
| `422` | `crate_json` is either missing, empty or invalid, or the validation could not run (an `error` result) |

## Validate a stored RO-Crate

`POST /v1/ro_crates/{crate_id}/validation`

> This queues validation of an RO-Crate held in the object store. The RO-Crate is resolved first, so a missing or ambiguous crate ID may be reported; the validation itself runs on a worker. 

Author: ```suggestion
This queues validation of an RO-Crate held in the object store. The RO-Crate is resolved first, so a missing or ambiguous crate ID may be reported immediately; the validation process itself runs asynchronously on a worker. 
```

!!! note
    See [Crate IDs](#crate-ids) for how `{crate_id}` maps to objects in the bucket.

| Field | Required | Description |
|-------|----------|-------------|
| `profile_name` | no | Profile to validate against; defaults to `ro-crate-1.1` when omitted |
| `webhook_url` | no | URL that receives the result when validation finishes |

```bash
curl -X POST http://localhost:5001/v1/ro_crates/my-dataset-2026/validation -H 'Content-Type: application/json' -d '{"profile_name": "ro-crate-1.2"}'
```

| Code | Meaning |
|------|---------|
| `202` | Queued; the body is `{"message": "Validation in progress"}` |
| `400` | Invalid Crate ID |

> | `404` | No RO-Crate at the expected keys, or storage mode is not enabled |

Author: ```suggestion
| `404` | Either storage mode is not enabled, or there is no RO-Crate at the location defined by the given Crate ID |
```

| `409` | Both a zip and a directory exist for this Crate ID |
| `422` | Request body invalid |
| `503` | Object store unreachable |

## Fetch a validation result

`GET /v1/ro_crates/{crate_id}/validation`

This returns the stored result for an RO-Crate.

```bash
curl http://localhost:5001/v1/ro_crates/my-dataset-2026/validation
```

| Code | Meaning |
|------|---------|
| `200` | The stored result, including persisted `error` results |
| `400` | Invalid Crate ID |

> | `404` | No result stored for this Crate ID yet |

Author: ```suggestion
| `404` | No result stored for this Crate ID |
```

## Validation results

Every validation produces a result object:

```json
{
  "status": "invalid",
  "profile": "ro-crate-1.2",
  "created_at": "2026-07-22T10:30:00+00:00",
  "detail": {}
}
```

An RO-Crate's `status` can be:

| `status` | Meaning |
|----------|---------|
| `valid` | The RO-Crate conforms to the profile |

> | `invalid` | Validated, but with conformance issues listed in `detail` |

Author: ```suggestion
| `invalid` | The RO-Crate does not conform to the profile, issues listed in `detail` field |
```

> | `error` | The validation could not run; the reason is in an `error` field instead of `detail` |

Author: ```suggestion
| `error` | The validation could not run; the reason is in an `error` field. No `detail` field is provided |
```
